src/messages.rs

```diff
@@ -185,57 +193,93 @@ impl HopMessage {
 #[derive(Ord, PartialOrd, Eq, PartialEq, Clone, Hash, RustcEncodable, RustcDecodable)]
 pub struct SignedMessage {
     /// A request or response type message.
     content: RoutingMessage,
-    /// Claimed public ID of a node or client.
-    ///
-    /// For clients this is easily verifiable since their name is computed from the ID. For nodes it
-    /// needs to be confirmed by their `NodeManager`.
-    public_id: PublicId,
-    signature: sign::Signature,
+    /// The lists of the groups involved in routing this message, in chronological order.
+    grp_lists: Vec<GroupList>,
+    /// The IDs and signatures of the source authority's members.
+    signatures: BTreeMap<PublicId, sign::Signature>,
 }
 
 impl SignedMessage {
     /// Creates a `SignedMessage` with the given `content` and signed by the given `full_id`.
     pub fn new(content: RoutingMessage, full_id: &FullId) -> Result<SignedMessage, RoutingError> {
-        let bytes_to_sign = try!(serialise(&(&content, full_id.public_id())));
+        let sig = sign::sign_detached(&try!(serialise(&content)), full_id.signing_private_key());
         Ok(SignedMessage {
             content: content,
-            public_id: *full_id.public_id(),
-            signature: sign::sign_detached(&bytes_to_sign, full_id.signing_private_key()),
+            grp_lists: Vec::new(),
+            signatures: iter::once((*full_id.public_id(), sig)).collect(),
         })
     }
 
-    /// Confirms the signature against the claimed public ID.
+    /// Confirms the signatures.
     pub fn check_integrity(&self) -> Result<(), RoutingError> {
-        let signed_bytes = try!(serialise(&(&self.content, &self.public_id)));
-        if sign::verify_detached(&self.signature,
-                                 &signed_bytes,
-                                 self.public_id().signing_public_key()) {
-            Ok(())
-        } else {
-            Err(RoutingError::FailedSignature)
+        let signed_bytes = try!(serialise(&self.content));
+        for (pub_id, sig) in &self.signatures {
+            if !sign::verify_detached(sig, &signed_bytes, pub_id.signing_public_key()) {
+                return Err(RoutingError::FailedSignature);
+            }
+        }
+        if let Authority::Client { ref client_key, .. } = self.content.src {
+            if let Some(pub_id) = self.signatures.keys().next() {
+                if self.signatures.len() == 1 && pub_id.signing_public_key() == client_key {
+                    return Ok(());
+                }
+            }
+            return Err(RoutingError::FailedSignature);
+        }
+        Ok(())
+    }
+    /// Returns whether the message is signed by the given public ID.
+    pub fn signed_by(&self, pub_id: &PublicId) -> bool {
+        self.signatures.contains_key(pub_id)
+    }
+
+    /// Appends the group list to the message, and removes all invalid signatures.
+    pub fn add_group_list(&mut self, group_list: GroupList) {
```

Comment:
Ah - I missed that `add_group_list()` was doing this.

The problem with checking is that we can receive the signatures and the actual `SignedMessage` in random order, so when we receive a signature we maybe haven't added the group list yet, or vice-versa.

I suppose we could change the behaviour to fix the first case (e.g. have ability to call `add_group_list()` multiple times for same group and only actually do the addition once - then we could call this when handling each incoming signature, or just pass the group list as an arg of `add_signature/s()`), but I don't see a simple way round the "vice-versa" scenario.

I guess I'd be inclined to try and fix this in `remove_invalid_signatures()` since when that's called we've definitely added the group and enough signatures.  So I'd remove the check from here and instead add it to the existing check in `remove_invalid_signatures()`.  Does that sound OK?